tensorrt_llm/_torch/debug/debug_hook.py

```diff
@@ -0,0 +1,385 @@
+import inspect
+import os
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Optional
+
+import torch
+import torch.nn as nn
+
+import tensorrt_llm
+
+
+class DebuggerContext:
+    """
+    A context container which contains the running states, such as the layer structures,
+    log folder, hooks to run, is pre_forward or after forward, etc.
+
+    Arguments:
+        dest_folder: str
+                The working directory set to debug context to set where the hook dumped data/info.
+    """
+
+    def __init__(self, dest_folder: str = None):
+        self.pre_forward_actions = {}
+        self.after_forward_actions = {}
+
+        self.layer_names = []
+        self.layer_inner_counter = []
+
+        self.module_forward_hook_handle = None
+        self.module_forward_pre_hook_handle = None
+
+        self.forward_hook_handles = {}  # module to handlers
+        self.forward_pre_hook_handles = {}
+        self.log_folder = dest_folder
+        self.is_pre_forward = True
+        self.is_log_folder_inited: bool = False
+        self._init_log_folder()
+
+    def _init_log_folder(self):
+        if self.is_log_folder_inited:
+            return
+
+        if self.log_folder is None:
+            pwd = os.getcwd()
+            self.log_folder = os.path.join(pwd, "data_dump")
+
+        rank = tensorrt_llm.mpi_rank()
+
+        p = Path(self.log_folder) / f"rank{rank}"
+        self.log_folder = p.absolute()
+        p.mkdir(parents=True, exist_ok=True)
+        self.is_log_folder_inited = True
+
+    def get_log_folder(self):
+        return self.log_folder
+
+    def check_in_pre_forward(self):
+        return self.is_pre_forward
+
+    def mark_in_pre_forward(self, is_pre_forward):
+        self.is_pre_forward = is_pre_forward
+
+    def clear_state(self):
+        self.pre_forward_actions.clear()
+        self.after_forward_actions.clear()
+        self.layer_names.clear()
+        self.layer_inner_counter.clear()
+
+        if self.module_forward_hook_handle is not None:
+            self.module_forward_hook_handle.remove()
+        if self.module_forward_pre_hook_handle is not None:
+            self.module_forward_pre_hook_handle.remove()
+
+        self.module_forward_hook_handle = None
+        self.module_forward_pre_hook_handle = None
+
+        for _, handler in self.forward_hook_handles.items():
+            handler.remove()
+
+        for _, handler in self.forward_pre_hook_handles.items():
+            handler.remove()
+
+        self.forward_hook_handles.clear()
+        self.forward_pre_hook_handles.clear()
+
+    def register_pre_forward_action(self, filter, action):
+        self.pre_forward_actions[filter] = action
```

Comment:
To make a filter be the key of a dictionary means the `Filter` class must be hashable. Do we need to look up an action giving a filter? If no, I would suggest we use tuple to get ride of such  constraint.

```python
self.pre_forward_actions = []
self.pre_forward_actions.append((filter, action))
```